qui/clipboard.py

```diff
@@ -122,31 +123,140 @@ def clipboard_formatted_size() -> str:
+                        None,  # parent_menu_item
+                        None,  # func
+                        None,  # data
+                        event.button,  # button
+                        Gtk.get_current_event_time())  # activate_time
+
+    def update_clipboard_contents(self, vm=None, size=0, message=None):
+        if not vm or not size:
+            self.clipboard_label.set_markup("<i>Qubes clipboard is empty</i>")
+            self.icon.set_from_icon_name("edit-copy")
+            # todo the icon should be empty and full depending on state
+
+        else:
+            self.clipboard_label.set_markup(
+                "<i>Qubes clipboard contents: {0} from "
+                "<b>{1}</b></i>".format(size, vm))
+            self.icon.set_from_icon_name("edit-copy")
+
+        if message:
+            self.notify(message)
+
+    def prepare_menu(self):
+        self.menu = Gtk.Menu()
+
+        title_label = Gtk.Label(xalign=0)
+        title_label.set_markup("<b>Current clipboard</b>")
+        title_item = Gtk.MenuItem()
+        title_item.set_sensitive(False)
+        title_item.add(title_label)
+        self.menu.append(title_item)
+
+        clipboard_content_item = Gtk.MenuItem()
+        clipboard_content_item.set_sensitive(False)
+        clipboard_content_item.add(self.clipboard_label)
+        self.update_clipboard_contents()
+        self.menu.append(clipboard_content_item)
+
+        self.menu.append(Gtk.SeparatorMenuItem())
+
+        help_label = Gtk.Label(xalign=0)
+        help_label.set_markup("<i>Use <b>Ctrl+Shift+V</b> to paste, "
+                              "<b>Ctrl+Shift+C</b> to copy.</i>")
+        help_item = Gtk.MenuItem()
+        help_item.set_margin_left(10)
+        help_item.set_sensitive(False)
+        help_item.add(help_label)
+        self.menu.append(help_item)
+
+        self.menu.append(Gtk.SeparatorMenuItem())
+
+        dom0_item = Gtk.MenuItem("Copy dom0 clipboard")
+        dom0_item.connect('activate', self.copy_dom0_clipboard)
+        self.menu.append(dom0_item)
+
+    def copy_dom0_clipboard(self, *_args, **_kwargs):
+        clipboard = Gtk.Clipboard.get(Gdk.SELECTION_CLIPBOARD)
+        text = clipboard.wait_for_text()
+
+        if not text:
+            self.notify("dom0 clipboard is empty!")
+            return
+
+        try:
+            fd = os.open(APPVIEWER_LOCK, os.O_RDWR | os.O_CREAT, 0o0666)
+        except Exception:  # pylint: disable=broad-except
+            self.notify("Error while accessing Qubes clipboard!")
+        else:
+            try:
+                fcntl.flock(fd, fcntl.LOCK_EX)
+            except Exception:  # pylint: disable=broad-except
+                self.notify("Error while locking Qubes clipboard!")
+            else:
+                try:
+                    with open(DATA, "w") as contents:
+                        contents.write(text)
+                    with open(FROM, "w") as source:
+                        source.write("dom0")
```

Comment:
You should also update xevent file with current X11 time. I see there is `gdk_x11_get_server_time()` in C, I think in pygtk it is `gtk.gdk.x11_get_server_time()`. Some hints:
https://developer.gnome.org/gdk3/stable/gdk3-X-Window-System-Interaction.html#gdk-x11-get-server-time
https://stackoverflow.com/a/5441492

Reply:
added